[PIPE-9694] Update serial groups docs for dangling unlock behavior change (#10626)
circleci/circleci-docs: docs/guides/modules/orchestrate/pages/controlling-serial-execution-across-your-organization.adoc

* First, the job enters a "queued" state.
* Once a job starts (reaches its place in the queue), it transitions to a "running" state.
* Upon completion (whether successful, failed, canceled, unauthorized, or not run), the lock is released and the next job in the queue starts. The queue continues to process jobs even if some jobs fail.

> Releasing the lock is separate from whether downstream workflow jobs run. A downstream job that uses `requires` to depend on a serial group job or job group invocation does not run just because the lock was released. By default, it only runs if that job succeeded. For a job group, this means all of the group's leaf jobs must have succeeded. You can change this default and react to other statuses instead of success by specifying the status alongside the dependency in the downstream job's `requires`. See <<how-job-group-expansion-works,How job group expansion works>>.

Reviewer: ```suggestion
Releasing the lock is separate from whether downstream workflow jobs run. A downstream job that uses `requires` to depend on a serial group job or job group invocation does not run just because the lock was released. By default, it only runs if that job succeeded. For a job group, this means all of the group's leaf jobs must have succeeded. You can change this default and react to other statuses instead of success by specifying the status (`success` and/or `not_run`) alongside the dependency in the downstream job's `requires`. See <<how-job-group-expansion-works,How job group expansion works>>.
```

When CircleCI processes a job group invocation, it inlines all jobs from the group definition and adds synthetic boundary jobs at each end:

* *Entry job*: Inserted before the root jobs (jobs in the group with no internal `requires` dependencies). With `serial-group`, this is a serial group start job that acquires the lock. Without it, this is a no-op group-start job.
* *Exit job*: Inserted after the leaf jobs (jobs in the group that no other group job depends on). This is a no-op job that only runs if the leaf jobs succeed.
** With `serial-group`, an additional serial group end job (unlock) is inserted after the group-end job. It runs regardless of the group-end job's outcome, so the lock is always released.

> The group invocation's `requires` and branch/tag filters are applied to the entry job. Jobs in the workflow that depend on the group invocation using `requires` are rewired to depend on the group-end job. By default, this means downstream jobs only run if all leaf jobs in the group succeeded, even though the lock is released regardless of outcome. To run a downstream job on a different outcome, specify the status alongside the dependency in `requires`.

Reviewer: ```suggestion
The group invocation's `requires` and branch/tag filters are applied to the entry job. Jobs in the workflow that depend on the group invocation using `requires` are rewired to depend on the group-end job. By default, this means downstream jobs only run if all leaf jobs in the group succeeded, even though the lock is released regardless of outcome. To run a downstream job on a different outcome, specify the status (`success` and/or `not_run`) alongside the dependency in `requires`.
```

pass:[<!-- vale off -->]

> If a downstream job configured to run on failure does not trigger after a job group fails, it may be waiting on the group's group-end job rather than the individual job that failed. Downstream jobs that depend on a job group invocation observe the group-end job's status (success only), not the status of individual jobs inside the group.

Reviewer: it says `(success only)` but it can also be `not_run`, right?

Author: Ah this can be confusing on a second read. I meant that group-end relies on success only of it's upstream jobs. 

Maybe something like this instead?

```suggestion
If a downstream job configured to run on failure does not trigger after a job group fails, this is because it depends on the group's group-end job. Downstream jobs that depend on a job group invocation react to the group-end job's status, not the status of any one job within the group. If a job within the job-group fails, group-end will have a status of `not_run`, rather than `failure`.
```

Author: This also highlights the gap we have that doesn't allow users to require a failing job group

Reviewer: this sounds good
do you mean that not_run is not a status one can add in requires?

Author: Yeah that's right
